### ISSUE TITLE
Deprecation warning for invalid base64 should only be issued if fixed base64 is valid

### ISSUE THREAD

**Issue by jeremyevans** [jwt/ruby-jwt#595]
Currently, passing anything that isn't strict urlsafe base64 will result in a deprecation warning.  Since this is usually passed user input, I'm not sure how one is supposed to avoid the deprecation warning, other than turning it off completely.  Ideally, the deprecation warning should only be emitted if the provided base64 is not strict urlsafe base64, but still results in a valid JWT using the loose base64 parsing.  The way it is currently implemented, you get the warning even if the base64 does not result in a JWT:

```ruby
# Does not and should not warn:
JWT.decode(JWT.encode('a', 'b'), 'b')

# Should warn and does:
JWT.decode(JWT.encode('a', 'b') + ' ', 'b')

# Warns but should not, as it raises JWT::VerificationError:
JWT.decode(JWT.encode('a', 'b') + ' 9', 'b')
```

If `JWT.decode` would raise, it should not issue a deprecation warning for invalid base64.

I realize this change requires refactoring how the deprecation warning is issued, since you would need to delay emission of the warning until after attempting parsing.

**Comment by anakinj** [comment]
Very valid point, user based deprecation warnings are hard :).

I have a pretty simple idea how this could be handled, a kinda conditional deprecation message concept.

**Comment by anakinj** [comment]
Fix released in 2.8.2
